toolchain: gcc: fix GCC < 11.0.0 build with -Wundef

idruzhinin opened the pull request:
toolchain: gcc: fix GCC < 11.0.0 build with -Wundef

When building using GCC version < 11.0.0, TOOLCHAIN_CLANG_VERSION is not
yet defined in gcc.h. Nevertheless, it's still needed due to reuse of
gcc.h in llvm.h.

Guard access to TOOLCHAIN_CLANG_VERSION to fix that.

dcpleung reviewed (changes requested):
`llvm.h` includes `gcc.h`. If you wants to remove `TOOLCHAIN_CLANG_VERSION` from `gcc.h`, you will need to define `FUNC_NO_STACK_PROTECTOR` under `llvm.h`, and make `gcc.h` to NOT re-define it.

idruzhinin commented:
Alternatively, the diff below also work I think, if that's look nicer:
```
diff --git a/include/zephyr/toolchain/gcc.h b/include/zephyr/toolchain/gcc.h
index f81cb0a309..089ee4c9ca 100644
--- a/include/zephyr/toolchain/gcc.h
+++ b/include/zephyr/toolchain/gcc.h
@@ -633,7 +633,8 @@ do {                                                                    \
  *
  * @note Only supported for GCC >= 11.0.0 or Clang >= 7.
  */
-#if (TOOLCHAIN_GCC_VERSION >= 110000) || (TOOLCHAIN_CLANG_VERSION >= 70000)
+#if (TOOLCHAIN_GCC_VERSION >= 110000) || \
+    (defined(TOOLCHAIN_CLANG_VERSION) && (TOOLCHAIN_CLANG_VERSION >= 70000))
 #define FUNC_NO_STACK_PROTECTOR __attribute__((no_stack_protector))
 #else
 #define FUNC_NO_STACK_PROTECTOR
```

kartben commented:
Ping @idruzhinin :)

idruzhinin commented:
sorry, was tide up recently and forgot I had that one in queue
could you specify which version of the patch you'd prefer - the one in PR or alternative (from the comment above)?

idruzhinin committed:
toolchain: gcc: fix GCC < 11.0.0 build with -Wundef

When building using GCC version < 11.0.0, TOOLCHAIN_CLANG_VERSION is not
yet defined in gcc.h. Nevertheless, it's still needed due to reuse of
gcc.h in llvm.h.

Guard access to TOOLCHAIN_CLANG_VERSION to fix that.

Signed-off-by: Igor Druzhinin <igord@xilinx.com>

kartben commented:
@tejlmand @stephanosio can you please have a look?
